### PR TITLE
add app.isPackaged

### PR DESCRIPTION
**Pull request by codebytere**
Resolves https://github.com/electron/electron/issues/7714.

First pass at adding an `isPackaged()` method to `app` in order to distinguish between development and production.

Todo:
- [x] specs
- [x] docs

/cc @zeke

### CODE REVIEW

**Comment by zeke** [comment]
For testing the `app.isPackaged() === true` scenario, we were thinking of writing a small app that does `process.stdout.write(String(app.isPackaged())`, packaging it with `electron-packager`, running it, and asserting the output. But perhaps there's a better way. Any ideas?

File: lib/browser/api/app.js [hunk 1]
```diff
@@ -39,6 +41,10 @@ Object.assign(app, {
   }
 })
 
+app.isPackaged = () => {
+  return ['app', 'app.asar'].some(p => fs.existsSync(path.join(process.resourcesPath, p)))
```
Comment (poiru) [inline lib/browser/api/app.js]: I think this could be done without a sync `fs` call by just comparing against `__dirname`.

Comment (codebytere) [reply lib/browser/api/app.js]: The above seems like a best case to me

**Comment by sindresorhus** [comment]
I think this would be better as a non-method property. It's a static check, so no need to execute it each time:

```js
app.isPackaged = (() => {
  const execFile = path.basename(process.execPath)
  if (process.platform === 'win32') {
    return execFile === 'electron.exe'
  }
  return execFile === 'electron'
})();
```

File: lib/browser/api/app.js [hunk 2]
```diff
@@ -39,6 +41,14 @@ Object.assign(app, {
   }
 })
 
+app.isPackaged = () => {
+  const execFile = path.basename(process.execPath)
+  if (process.platform === 'win32') {
+    return execFile === 'electron.exe'
```
Comment (MarshallOfSound) [inline lib/browser/api/app.js]: heh, `!==`

File: lib/browser/api/app.js [hunk 3]
```diff
@@ -39,6 +41,14 @@ Object.assign(app, {
...
+  }
+  return execFile === 'electron'
```
Comment (MarshallOfSound) [inline lib/browser/api/app.js]: heh, `!==`

File: lib/browser/api/app.js (hunk 2 above)
Comment (codebytere) [reply lib/browser/api/app.js]: lol whoops

**Comment by zeke** [comment]
Updated docs and tests to account for change from a function to a property. Still interested in ideas on how to test the `app.isPackaged === true` scenario.

**Comment by codebytere** [comment]
I've updated based on latest changes, but i'm not sure we should try to test the packaged scenario given that we would then have to rely on `electron-packager` which would introduce a cyclic dependency. If we test the positive scenario locally to success, i'm comfortable just testing the false scenario in specs and leaving the other out.

**Comment by zeke** [comment]
Thanks for making this happen @codebytere 🌟

**Comment by Mikaeru69** [comment]
Until now, to distinguish between development and production , I've been using
`const isPackaged = !process.defaultApp;`
from the main process.
I wonder whether the planned `app.isPackaged` will return a different result...

> process.defaultApp
> A Boolean. When app is started by being passed as parameter to the default app, this property is true in the main process, otherwise it is undefined.

**Comment by zeke** [comment]
Quoting from https://github.com/electron/electron/issues/7714


> Electron has a property called `process.defaultApp` which is set to `true` when running Electron with a target directory, e.g. `electron .`. There are at least two problems with this:
> 1. Only works in the main process. Rendered processes can't access this property by default.
> 2. The name `defaultApp` is not intuitive, and the [docs give no indication](http://electron.atom.io/docs/all/#processdefaultapp) of this property's practical use.

**Comment by Mikaeru69** [comment]
1. Well, in order to access these properties from a renderer process, one still needs to make use of remote...

`const isPackaged = require('electron').remote.process.defaultApp || false;`

`const isPackaged = require('electron').remote.app.isPackaged;`

2. Absolutely right!

**Comment by schontz** [comment]
What version of electron is this available in?

**Comment by schontz** [comment]
Nvm. Found it on the blog post, but it hasn't made its way to the docs yet.

**Comment by shiftkey** [comment]
@schontz https://electronjs.org/docs/api/app#appispackaged

**Comment by schontz** [comment]
Weird, it doesn't show up in search.

**Comment by zeke** [comment]
My bad, probably. @schontz can you file an issue on https://github.com/electron/algolia-indices please?

**Comment by schontz** [comment]
https://github.com/electron/algolia-indices/issues/25

**Pull request merged**